base/system/winlogon/sound.c

```diff
@@ -0,0 +1,325 @@
+/*
+ * PROJECT:     ReactOS Winlogon
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     System sound notifications
+ * COPYRIGHT:   Copyright 2011 Rafal Harabien <rafalh@reactos.org>
+ *              Copyright 2013 Eric Kohl <eric.kohl@reactos.org>
+ *              Copyright 2016 Hermès Bélusca-Maïto <hermes.belusca-maito@reactos.org>
+ *              Copyright 2021 Thamatip Chitpong <tangaming123456@outlook.com>
+ */
+
+/* INCLUDES *****************************************************************/
+
+#include "winlogon.h"
+
+#include <mmsystem.h>
+#include <winsvc.h>
+
+/* GLOBALS ******************************************************************/
+
+typedef struct _WL_PLAYSOUND_DATA
+{
+    WL_SYSTEM_SOUND Sound;
+    PWLSESSION Session;
+} WL_PLAYSOUND_DATA, *PWL_PLAYSOUND_DATA;
+
+/* FUNCTIONS ****************************************************************/
+
+static
+BOOL
+PlaySoundRoutine(
+    IN LPCWSTR lpFileName,
+    IN BOOL bLogon,
+    IN DWORD dwFlags)
+{
+    typedef BOOL (WINAPI *PFPLAYSOUNDW)(LPCWSTR, HMODULE, DWORD);
+    typedef UINT (WINAPI *PFWAVEOUTGETNUMDEVS)(VOID);
+
+    PFPLAYSOUNDW _PlaySoundW;
+    PFWAVEOUTGETNUMDEVS _waveOutGetNumDevs;
+    UINT uNumDevs;
+    HMODULE hLibrary;
+    BOOL bRet = FALSE;
+
+    hLibrary = LoadLibraryW(L"winmm.dll");
+
+    if (hLibrary)
+    {
+        _waveOutGetNumDevs = (PFWAVEOUTGETNUMDEVS)GetProcAddress(hLibrary, "waveOutGetNumDevs");
```

Comment:
It would be of better interest to load and get the needed functions once and for all, instead of everytime a sound is triggered.
However, the `_waveOutGetNumDevs` call below (to check for devices == 0 or not) needs to be done everytime.

Reply:
That's a good idea, but the problem is the function pointers have to be valid while the thread is running.